### ISSUE TITLE
[ERROR   ] Exception Require 20 byte binary sha, got '\xe6\x90', len = 2 occurred in file server update

### ISSUE THREAD

**Issue by PredatorVI** [saltstack/salt#35239]
Running on Ubuntu 14.04.5 with latest salt-master version 2016.3.2 packages.  It also happened recently with previous (newish) versions.  

The logs show the exception below on the salt-master during what appears to be a refresh of gitfs repos.  The result is that none of my SLS files are loaded and I'm dead in the water.

When I saw the initial issue I also saw warnings related to changing hash_type to sha256 so I tried changing the hash_type as well thinking that was related.  As part of the troubleshooting process I also upgrade to 2016.3.1 but the problem didn't go away, so I purged all salt packages and all unused dependent packages and did a clean install.  The problem went away for about a week, but yesterday it started happening again so I upgraded to 2016.3.2 (also tried a purge+clean install).

Here is the full exception: 
```
[ERROR   ] Exception Require 20 byte binary sha, got '\xe6\x90', len = 2 occurred in file server update
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/daemons/masterapi.py", line 270, in fileserver_update
    fileserver.update()
  File "/usr/lib/python2.7/dist-packages/salt/fileserver/__init__.py", line 453, in update
    self.servers[fstr]()
  File "/usr/lib/python2.7/dist-packages/salt/fileserver/gitfs.py", line 124, in update
    gitfs.update()
  File "/usr/lib/python2.7/dist-packages/salt/utils/gitfs.py", line 2218, in update
    self.find_file
  File "/usr/lib/python2.7/dist-packages/salt/fileserver/__init__.py", line 240, in reap_fileserver_cache_dir
    ret = find_func(filename, saltenv=saltenv)
  File "/usr/lib/python2.7/dist-packages/salt/utils/gitfs.py", line 2577, in find_file
    blob, blob_hexsha = repo.find_file(repo_path, tgt_env)
  File "/usr/lib/python2.7/dist-packages/salt/utils/gitfs.py", line 862, in find_file
    file_blob = tree / path
  File "/usr/lib/python2.7/dist-packages/git/objects/tree.py", line 172, in __div__
    item = tree[token]
  File "/usr/lib/python2.7/dist-packages/git/objects/tree.py", line 237, in __getitem__
    return self.__div__(item)
  File "/usr/lib/python2.7/dist-packages/git/objects/tree.py", line 188, in __div__
    return self._map_id_to_type[info[1] >> 12](self.repo, info[0], info[1], join_path(self.path, info[2]))
  File "/usr/lib/python2.7/dist-packages/git/objects/tree.py", line 133, in __init__
    super(Tree, self).__init__(repo, binsha, mode, path)
  File "/usr/lib/python2.7/dist-packages/git/objects/base.py", line 142, in __init__
    super(IndexObject, self).__init__(repo, binsha)
  File "/usr/lib/python2.7/dist-packages/git/objects/base.py", line 39, in __init__
    assert len(binsha) == 20, "Require 20 byte binary sha, got %r, len = %i" % (binsha, len(binsha))
AssertionError: Require 20 byte binary sha, got '\xe6\x90', len = 2
```

###/etc/salt/master (all commented lines removed):###
```
worker_threads: 50
timeout: 240
loop_interval: 30
ping_on_rotate: True
open_mode: True
auto_accept: True
external_auth:
  pam:
    go.salt:
      - .*
      - '@runner'
      - '@jobs'
      - '@wheel'
    jeff.vincent:
      - .*
      - '@runner'
      - '@wheel'
      - '@jobs'

auth.ldap.basedn: mybasedn
auth.ldap.binddn: mybinddn
auth.ldap.bindpw: blahblah
auth.ldap.filter: sAMAccountName={{ username }}
auth.ldap.server: myserver

state_output: full
hash_type: sha256
file_ignore_regex:
  - '/\.svn($|/)'
  - '/\.git($|/)'

fileserver_backend:
  - git

gitfs_remotes:
  - ssh://git@gitlab.wp.bluecoat.com/webpulse/salt-states.git
  - ssh://git@gitlab.wp.bluecoat.com/webpulse/salt-formula.git
  - ssh://git@gitlab.wp.bluecoat.com/webpulse/rabbitmq-formula.git
  - ssh://git@gitlab.wp.bluecoat.com/webpulse/salt-formula-freeipa.git

gitfs_ssl_verify: False
pillar_roots:
  base:
    - /srv/pillar/base
  dev:
    - /srv/pillar/dev
    - /srv/pillar/base
  acc:
    - /srv/pillar/acc
    - /srv/pillar/base
  int:
    - /srv/pillar/int
    - /srv/pillar/base
  stg:
    - /srv/pillar/stg
    - /srv/pillar/base
  prd-dc5:
    - /srv/pillar/prd-dc5
    - /srv/pillar/base
  prd-se1:
    - /srv/pillar/prd-se1
    - /srv/pillar/base

log_level: info
win_gitrepos:
  - 'https://github.com/saltstack/salt-winrepo.git'
```

Beyond running the server I don't know what specific things are causing the issue.
```
### Versions Report
Salt Version:
           Salt: 2016.3.2

Dependency Versions:
           cffi: 0.9.2
       cherrypy: 3.7.0
       dateutil: 1.5
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 0.9.1
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.2.3
      pycparser: 2.12
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.0.1
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-92-generic
         system: Linux
        version: Ubuntu 14.04 trusty
```

**Comment by Ch3LL** [comment]
@PredatorVI can you please try deleting your cache and restarting the master and seeing if that helps resolve the issue. `rm -rf /var/cache/salt/master/`

Also did you start seeing this behavior first on 2015.8 or only on 2016.3?

Also ping @terminalmage do you see anything obvious from this error message to help troubleshoot this. Thanks

**Comment by terminalmage** [comment]
There is nothing wrong with the Salt code here, all we're doing is obtaining the blob object from the relative path to the file, using ``file_blob = tree / path``.  This particular syntax has been in GitFS since its inception, years ago.

This is almost certainly an upstream bug in GitPython. According to your versions report, you have version 0.3.2 RC1 of GitPython installed, which is just over 5 years old. My recommendation is to upgrade GitPython to a newer release (version 2.0.8 was just released about a week ago), or trying out Pygit2 if you can find a new enough version (0.20.3 or newer) packaged for Ubuntu 14.04 LTS (probably unlikely).

As GitPython is pure Python (at least, last I checked), it should be fairly easy to install using ``pip``. You can ``apt-get remove`` the system package for ``python-git`` and ``python-gitdb`` and then install GitPython using ``pip install GitPython``. If pip is not installed, I believe it is packaged for Ubuntu as ``python-pip``.

**Comment by PredatorVI** [comment]
I do not believe I ever saw this error on 2015.8.  

The initial salt-master install is not that old and was a new install on Ubuntu 14.04 so the fact that the GitPython version is that old is disturbing.

I asked around the IRC channel about this issue and was told that PyGit2 on Ubuntu 14.04 was painful to install so it was suggested to move to Ubuntu 16.04 since there was an official package for pygit2 (python-pygit2), but salt-api is throwing an AssertionError so I can't test my process.  However, I don't see this exception in the logs.

I have successfully reverted back to Ubuntu 14.04 + Salt 2016.3 after clearing cache, removing system packages for python-git and python-gitdb and installing GitPython via pip.  I'll close this and reopen if needed.

**Comment by terminalmage** [comment]
GitPython went through about a 2 or 3-year period where it wasn't actively being developed, this is likely the reason for the older release being packaged in Ubuntu 14.04 LTS. There wasn't a newer one available at the time 14.04 was frozen, most likely.

As for why you didn't see the error, this could be as simple as GitPython not properly handling a given SHA, so this error could have appeared at literally any time.

GitPython wraps the git CLI, and in older versions relied quite heavily on very rigid ``assert`` statements that tried to confirm that the output of git commands matched what was expected. More recent GitPython versions seem to have improved its stability.
